### ISSUE TITLE
Invalid reference to object of type 'header' with name

### ISSUE THREAD

**Issue by mohammedelazzab** [p4lang/p4c#2431]
I wrote a simple p4 program that inserts a new header 'myTunnel', it compiles succesfully but when I run the generated json file using simple_switch (version 1.11.0.-unknown) I see the following error:
 
**Invalid reference to object of type 'header' with name 'myTunnel'** 

I included the "v1model.p4". The header is defined as follow: 

```
header myTunnel_t {
    port_t lport;
    ethertype_t  ethertype; 
}

struct headers {
    ethernet_t ethernet;
    myTunnel_t[MAX_SGMTS] myTunnel;
    ipv4_t     ipv4;
    tcp_t   tcp;
    udp_t   udp;
}
```

I tried the same thing with psa_switch and I have the same issue.
Can someone please explain me what could be the problem ?

**Comment by jafingerhut** [comment]
It would help us try to reproduce the problem if you could provide a complete P4 program that exhibits this problem.

**Comment by mohammedelazzab** [comment]
[myTunnel.p4.txt](https://github.com/p4lang/p4c/files/4771436/myTunnel.p4.txt)
Here is the p4 program.

**Comment by mbudiu-vmw** [comment]
I am getting several errors when trying to compile your program, so I cannot get to the execution phase at all. 

```
issue2341-bmv2.p4(152): [--Wwarn=deprecated] warning: mark_to_drop: Using deprecated feature mark_to_drop. Please use mark_to_drop(standard_metadata) instead.
        mark_to_drop();
        ^^^^^^^^^^^^
/home/mbudiu/git/p4c/p4include/v1model.p4(415)
extern void mark_to_drop();
            ^^^^^^^^^^^^
issue2341-bmv2.p4(97): [--Wwarn=uninitialized] warning: hdr.myTunners.next: reading uninitialized value
        transition select(hdr.myTunners.next.ethertype) {
                          ^^^^^^^^^^^^^^^^^^
issue2341-bmv2.p4(22): [--Werror=target-error] error: header myTunner_t: Found header with fields totaling 25 bits.  BMv2 target only supports headers with fields totaling a multiple of 8 bits.
header myTunner_t {
       ^^^^^^^^^^
issue2341-bmv2.p4(97): [--Werror=uninitialized] error: hdr.myTunners.next uninitialized: next field read
        transition select(hdr.myTunners.next.ethertype) {
                          ^^^^^^^^^^^^^^^^^^
issue2341-bmv2.p4(152): [--Werror=target-error] error: Expected 1 argument for mark_to_drop. Are you using an up-to-date v1model.p4?
        mark_to_drop();
        ^^^^^^^^^^^^^^
```

**Comment by mohammedelazzab** [comment]
It compiles for me properly. I am using the p4c compiler from the  cslev/p4c container:
```
sudo docker run -it -d --privileged -v $(pwd):/root/p4src --name p4c-cr cslev/p4c bash
sudo docker exec -w /root/p4src -it p4c-cr p4c --target bmv2 --arch v1model --std p4-16 $1
sudo docker rm p4c-cr -f
```
The p4c version is 0.5
the container is downloadable from dockerhub under cslev/p4c:latest

**Comment by mbudiu-vmw** [comment]
The current p4c version in this repository is 1.2.
You should try with a newer version of the compiler.

**Comment by jafingerhut** [comment]
I downloaded your myTunnel.p4 program attached to an earlier comment, and made small edits to it, attached:

+ one to update mark_to_drop() for a change made to the v1model architecture about a year ago
+ change a header definition to contain a multiple of 8 bits, a restriction added to the v1model implementation also about a year ago
+ correct the parser so that the select expression uses .last instead of .next, because those refer to the same thing after doing an extract() on a .next of a header stack, after the extract.

and the program compiles with the latest version of p4c with no errors or warnings.  I did not do any testing on it.

It appears that whatever issue you are hitting in that older version of p4c has been fixed before now.

[myTunnel-updated.p4.txt](https://github.com/p4lang/p4c/files/4773498/myTunnel-updated.p4.txt)
